providers/implementations/keymgmt/ec_kmgmt.c

```diff
@@ -0,0 +1,253 @@
+
+    if (curve_nid == NID_undef) {
+        /* explicit parameters */
+
+        /*
+         * TODO(3.0): should we support explicit parameters curves?
+         */
+        return 0;
+    } else {
+        /* named curve */
+        const char *curve_name = NULL;
+
+        if ((curve_name = OBJ_nid2sn(curve_nid)) == NULL)
+            return 0;
+
+        if (!ossl_param_bld_push_utf8_string(tmpl, OSSL_PKEY_PARAM_EC_NAME, curve_name, 0))
+            return 0;
+    }
+
+    return 1;
+}
+
+static ossl_inline
+EC_KEY *params_to_key(const OSSL_PARAM params[])
+{
+    const OSSL_PARAM *param_priv_key, *param_pub_key;
+    BIGNUM *priv_key = NULL;
+    char *pub_key = NULL;
+    EC_GROUP *ecg = NULL;
+    EC_POINT *pub_point = NULL;
+    EC_KEY *ret = NULL;
+    int ok = 0;
+
+    if ((ecg = params_to_domparams(params)) == NULL)
+        return NULL;
+
+    param_priv_key =
+        OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_EC_PRIV_KEY);
+    param_pub_key =
+        OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_EC_PUB_KEY);
+
+    /*
+     * We want to have at least a public key either way, so we end up
+     * requiring it unconditionally.
+     */
+    if (param_pub_key == NULL)
+        goto err;
+
+    if ((param_priv_key != NULL
+                && !OSSL_PARAM_get_BN(param_priv_key, &priv_key))
+            || !OSSL_PARAM_get_utf8_string(param_pub_key, &pub_key, 0))
+        goto err;
+
+    if (!EC_POINT_hex2point(ecg, pub_key, pub_point, NULL))
+        goto err;
+
+    if ((ret = EC_KEY_new()) == NULL
+            || !EC_KEY_set_group(ret, ecg)
+            || !EC_KEY_set_private_key(ret, priv_key)
+            || !EC_KEY_set_public_key(ret, pub_point))
+        goto err;
+
+    ok = 1;
+
+ err:
+    BN_clear_free(priv_key);
+    OPENSSL_free(pub_key);
+    EC_POINT_free(pub_point);
+    EC_GROUP_free(ecg);
+    if (!ok) {
+        EC_KEY_free(ret);
+        ret = NULL;
+    }
+    return ret;
+}
+
+static ossl_inline
+int key_to_params(const EC_KEY *eckey, OSSL_PARAM_BLD *tmpl)
+{
+    const BIGNUM *priv_key = NULL;
+    const EC_POINT *pub_point = NULL;
+    const EC_GROUP *ecg = NULL;
+    char *pub_key = NULL;
+    int ret = 0;
+
+    if (eckey == NULL)
+        return 0;
+
+    ecg = EC_KEY_get0_group(eckey);
+    priv_key = EC_KEY_get0_private_key(eckey);
+    pub_point = EC_KEY_get0_public_key(eckey);
+
+    /* group and public_key must be present, priv_key is optional */
+    if (ecg == NULL || pub_point == NULL)
+        return 0;
+#define PNT_FORM POINT_CONVERSION_UNCOMPRESSED
+    if ((pub_key = EC_POINT_point2hex(ecg, pub_point, PNT_FORM, NULL)) == NULL)
+        return 0;
+
+    if (!domparams_to_params(ecg, tmpl))
+        goto err;
+    if (priv_key != NULL
+            && !ossl_param_bld_push_BN(tmpl,
+                                       OSSL_PKEY_PARAM_EC_PRIV_KEY,
+                                       priv_key))
+        goto err;
+    if (!ossl_param_bld_push_utf8_string(tmpl,
```

Comment:
I'm trying to decide whether it is better to pass the point in encoded form as the parameter. Or whether passing the x and y separately would be better.

Reply:
I debated the same, if encoded is supported by the FIPS module I would probably prefer the encoded format as I expect it would improve interoperability with future external providers, as the encoding format is coming from an external standard.

Reply:
Well, its not difficult to decode this format so I think it should be fine in the FIPS module.

Reply:
What would be the downside of passing x and y ?

Reply:
> What would be the downside of passing x and y ?

@levitte the downside is somewhat related to the discussion in https://github.com/openssl/openssl/issues/10366 : there, when parsing, we allow zero padded hexstrings that do not conform to the standard to be tolerated and interpreted as valid.

Why should we use a custom representation in the provider PARAMS "API" when a standard interoperable definition of how to represent a point as an hexstring exists?

Reply:
Oh, NOW I notice it...  you're passing back the public key in hex string form?  I don't find that acceptable, please pass it unmodified as an octet string.  Any serialization to hex form should be done elsewhere (such as a serializer for EC keys)

Reply:
Uhm.. I thought I already converted all the point2hex to point2oct but this must have slipped through... I am though worried that this is not causing failures in the tests..

Reply:
In the upcoming push this has been fixed using `EC_POINT_point2buf` to use an octet string rather than a utf8 string.